common/setups/rasr/util/decode.py

```diff
@@ -0,0 +1,163 @@
+__all__ = [
+    "DevRecognitionParameters",
+    "TestRecognitionParameters",
+    "RecognitionParameters",
+    "PriorPath",
+    "SearchJobArgs",
+    "Lattice2CtmArgs",
+    "StmArgs",
+    "ScliteScorerArgs",
+    "OptimizeJobArgs",
+]
+
+from dataclasses import dataclass
+from itertools import product
+from typing import Any, Dict, List, Optional, Tuple, TypedDict, Union
+
+from sisyphus import tk
+
+import i6_core.rasr as rasr
+
+from ..config.am_config import Tdp
+
+
+class DevRecognitionParameters:
+    def __init__(
+        self,
+        *,
+        am_scales: List[float],
+        lm_scales: List[float],
+        prior_scales: List[float],
+        pronunciation_scales: Optional[List[float]],
+        tdp_scales: List[float],
+        speech_tdps: List[Tdp],
+        silence_tdps: List[Tdp],
+        nonspeech_tdps: Optional[List[Tdp]],
+        altas: Optional[List[float]],
+    ):
+        self.am_scales = am_scales
+        self.lm_scales = lm_scales
+        self.prior_scales = prior_scales
+        self.pronunciation_scales = pronunciation_scales if pronunciation_scales is not None else [1.0]
+        self.tdp_scales = tdp_scales
+        self.speech_tdps = speech_tdps
+        self.silence_tdps = silence_tdps
+        self.nonspeech_tdps = nonspeech_tdps if nonspeech_tdps is not None else [None]
+        self.altas = altas if altas is not None else [0.0]
+
+    def _get_iter(self):
+        return [
+            (am, lm, pri, pron, tdp, speech, sil, nonspeech, al)
```

Comment:
When doing grid search on the decoding parameters for a hybrid model, doing such a cartesian product does not make so much sense. One should first tune model-related scales such as `prior scale` and `tdp scale`. Then tdp values,. And only given the optimal values of the mentioned parameters one tunes the lm and pronunciation scales. Moreover, we should definitely consider the obligatory use of a high `altas` and small `beam` for the first two steps - only `lm scale` should not be tuned together with `altas`.

Reply:
You have the option to pass a List of RecogParams or RecogParams with lists where the cartesian product is taken from. We can always change the RecogParam class later on or extend or add more versions.. that is why I tried to put it all in the RecogParamClass